h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -160,6 +173,125 @@ public long getModelCount() {
+    private final Grid grid;
+    private final Lock parallelSearchGridLock = new ReentrantLock();
+
+    public ModelFeeder(HyperSpaceWalker.HyperSpaceIterator<MP> hyperspaceIterator, Grid grid) {
+      this.hyperspaceIterator = hyperspaceIterator;
+      this.grid = grid;
+    }
+
+    private void onModelSuccessfullyBuilt(final Model finishedModel, final ParallelModelBuilder parallelModelBuilder) {
+      try {
+        constructScoringInfo(finishedModel);
+        parallelSearchGridLock.lock();
+        grid.putModel(finishedModel._parms.checksum(IGNORED_FIELDS_PARAM_HASH), finishedModel._key);
+        _job.update(1);
+        attemptGridSave(grid);
+
+        grid.update(_job);
+      } finally {
+        parallelSearchGridLock.unlock();
+      }
+      attemptBuildNextModel(parallelModelBuilder, finishedModel);
+    }
+
+    private void onModelBuildFailed(ParallelModelBuilder.ModelBuildFailure modelBuildFailure, ParallelModelBuilder parallelModelBuilder) {
+      parallelSearchGridLock.lock();
+      try {
+        grid.appendFailedModelParameters(null, modelBuildFailure.getParameters(), modelBuildFailure.getThrowable());
+      }finally {
+        parallelSearchGridLock.unlock();
+      }
+      attemptBuildNextModel(parallelModelBuilder, null);
+    }
+
+    private void attemptBuildNextModel(final ParallelModelBuilder parallelModelBuilder, final Model previousModel) {
+      // Attempt to train next model
+      final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
+      if (nextModelParams != null && isThereEnoughTime() && !_job.stop_requested()) {
+        parallelModelBuilder.run(ModelBuilder.make(nextModelParams));
+      } else {
+        parallelModelBuilder.noMoreModels();
+      }
+    }
+
+    private void constructScoringInfo(final Model model) {
+      ScoringInfo scoringInfo = new ScoringInfo();
+      scoringInfo.time_stamp_ms = System.currentTimeMillis();
+
+      model.fillScoringInfo(scoringInfo);
+      grid.setScoringInfos(ScoringInfo.prependScoringInfo(scoringInfo, grid.getScoringInfos()));
+      ScoringInfo.sort(grid.getScoringInfos(), _hyperSpaceWalker.search_criteria().stopping_metric());
+    }
+
+    private boolean isThereEnoughTime() {
+      final boolean enoughTime = hyperspaceIterator.max_runtime_secs() > 0 && hyperspaceIterator.time_remaining_secs() > 0;
+
+      if (!enoughTime) {
+        Log.info("Grid max_runtime_secs of " + hyperspaceIterator.max_runtime_secs() + " secs has expired; stopping early.");
+      }
+
+      return enoughTime;
+    }
+
+    private MP getNextModelParams(final HyperSpaceWalker.HyperSpaceIterator<MP> hyperSpaceWalker, final Model model, final Grid grid){
+      MP params = null;
+
+      while (params == null) {
+        if (hyperSpaceWalker.hasNext(Optional.ofNullable(model))) {
+          params = hyperSpaceWalker.nextModelParameters(Optional.ofNullable(model));
```

Comment:
In legacy
`public MP nextModelParameters(Optional<Model> previousModel)`
`public boolean hasNext(Optional<Model> previousModel)`
you're allowed to pass the previous model as if it was enough to decide about the next model params, or if there should be a next model. 
But this makes sense only in sequential mode. 
In parallel mode, the `previousModel` can be any of the models that was started until now. Only the Grid itself should be the source of truth to decide what's next.
Btw, none of the methods listed above is currently using this parameter, which is already suspicious... I would simply remove it.

Reply:
The contract is set, the `model` argument is optional and with Java 8, we're only making it more obvious. The API does not state the model, if provided, is the previous model time-wise. 

There might be strategies based on previous model finished I can imagine, based on model's metrics for example. With model, parameters used to build that model are available. This implies the hyperspace walker knows which part of the hyperspace to avoid/focus on, based on model performance (as one example). Such strategies are usable even in parallel regime. Therefore, I'd not close our way to obtaining the information about the model built, even if it is currently unused.

I'd discuss my changes to the API as a whole. If this is part of Java API, then we might as well want to keep those walkers as they are right now, not adding the Optional, nor moving them to a separate package.

Reply:
This belongs to the Walkers, I'm not sure we can consider those as the Java API (at least not from a user perspective), and I doubt that there are custom implementations of those walkers (although possible why not..., we could always find a way to make the change backwards-compatible).

I agree that there might be strategies that would rely on built models to decide about next step.
My point is just that (and especially in parallel mode), the last built model is non significant, the strategy should rely on all previously built models... there's no reason to limit this to only one model.
To me it just shows that the legacy API was designed with `sequential` mode in mind.

Reply:
To be clear, what I had in mind was early stopping in parallel mode.
If it were implemented in the random walker `hasNext` instead of the way it is currently done, then you would have it for free in parallel mode too... somehow. But for this you need to pass the full grid, not only the last model.

Sure we can always imagine strategies that would rely only on some characteristics of one single model, but I can't see any interesting one tbh.

Reply:
The stopEarly method acccepts "last model" and has access to all the previous scoring infos of models built in time. This contract I believe can be honored even in parallel mode.

Reply:
But I can also see the potential problems with that. The APIs are vaguely defined, but I don't have the balls to refactor them right now. Most of those are truly user-facing and there'd be a battle far beyond scope of this PR. For current walkers, it will work just fine. For future walkers ... not so sure.